onnxruntime/core/providers/cpu/math/einsum_auxiliary_ops.cc

```diff
@@ -0,0 +1,296 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "einsum_auxiliary_ops.h"
+
+namespace onnxruntime {
+
+namespace EinsumOp {
+
+Tensor Transpose(const Tensor& input, const std::vector<size_t>& permutation, AllocatorPtr allocator) {
+  const auto& input_dims = input.Shape().GetDims();
+  auto input_rank = input_dims.size();
+
+  ORT_ENFORCE(input_rank == permutation.size(), "Length of permutation must match the rank of the input to be permutated");
+
+  std::vector<int64_t> output_dims;
+  output_dims.reserve(input_rank);
+
+  for (const auto& dim : permutation) {
+    output_dims.push_back(input_dims.at(dim));
+  }
+
+  // Pass in allocator as that will be used as an allocator deleter by the framework
+  // and it will de-allocate the memory for this intermediate tensor when it goes out of scope
+  Tensor output(input.DataType(), output_dims, allocator);
+
+  TransposeBase::DoTranspose(permutation, input, output);
+
+  return output;
+}
+
+template <typename T>
+Tensor MatMul(const Tensor& input_1, const Tensor& input_2, AllocatorPtr allocator, concurrency::ThreadPool* tp) {
+  const auto& input1_dims = input_1.Shape().GetDims();
+  const auto& input2_dims = input_2.Shape().GetDims();
+
+  // Sanity checks before the actual MatMul
+  ORT_ENFORCE(input_1.DataType() == input_2.DataType(), "Data types of the inputs must match for MatMul");
+  ORT_ENFORCE(input1_dims.size() == 3 && input2_dims.size() == 3, "Only 1 batch dimension is allowed for MatMul");
+  ORT_ENFORCE(input1_dims[0] == input2_dims[0], "Batch dimension should match for MatMul;");
+  ORT_ENFORCE(input1_dims[2] == input2_dims[1], "Incompatible matrix dimensions for matMul");
+
+  size_t batches = static_cast<size_t>(input1_dims[0]);
+  size_t M = static_cast<size_t>(input1_dims[1]);
+  size_t K = static_cast<size_t>(input1_dims[2]);
+  size_t N = static_cast<size_t>(input2_dims[2]);
+
+  size_t left_offset = M * K;
+  size_t right_offset = K * N;
+  size_t output_offset = M * N;
+
+  std::vector<int64_t> output_dims;
+  output_dims.reserve(3);
+  output_dims.push_back(static_cast<int64_t>(batches));
+  output_dims.push_back(static_cast<int64_t>(M));
+  output_dims.push_back(static_cast<int64_t>(N));
+
+  // Pass in allocator as that will be used as an allocator deleter by the framework
+  // and it will de-allocate the memory for this intermediate tensor when it goes out of scope
+  Tensor output(input_1.DataType(), output_dims, allocator);
+
+  const T* input_1_data = input_1.template Data<T>();
+  const T* input_2_data = input_2.template Data<T>();
+  T* output_data = output.template MutableData<T>();
+
+  // Process each batch
+  for (size_t i = 0; i < batches; ++i) {
+    math::MatMul<T>(
+        static_cast<int>(M),
+        static_cast<int>(N),
+        static_cast<int>(K),
+        input_1_data + i * left_offset,
+        input_2_data + i * right_offset,
+        output_data + i * output_offset, tp);
+  }
```

Comment:
May be better to parallelize the batches vs the individual MatMul calls if there are multiple batches.

Reply:
Right now the general interface exposed in math.h for MatMul seems to parallelize a single MatMul (for MLAS enabled builds - the single MatMul operation is parallelized). Even our dedicated MatMul kernel seems to only parallelize within a single MatMul operation. I can add a TODO here that we can take action on later that detremines whether to parallelize on batches or within a single operation. WIll that be okay ? (I have a similar TODO in the Diagonal op as well)

Reply:
Added a TODO.  Please let me know if that make sense.

Reply:
may replace 'if we can' with 'when'

---
In reply to: [414889736](https://github.com/microsoft/onnxruntime/pull/3575#discussion_r414889736) [](ancestors = 414889736)